plugins/flytekit-pydantic/flytekitplugins/pydantic/deserialization.py

```diff
@@ -0,0 +1,148 @@
+        """
+        assert not cls.is_attached(), "can only be attached to one literal map at a time."
+        try:
+            cls.literal_store = literal_map.literals
+            yield
+        finally:
+            cls.literal_store = None
+
+    @classmethod
+    def contains(cls, item: commons.LiteralObjID) -> bool:
+        assert cls.is_attached(), "can only check for existence of a literal when attached to a literal map"
+        assert cls.literal_store is not None
+        return item in cls.literal_store
+
+    @classmethod
+    def is_attached(cls) -> bool:
+        return cls.literal_store is not None
+
+    @classmethod
+    def get_python_object(
+        cls, identifier: commons.LiteralObjID, expected_type: Type[PythonType]
+    ) -> Optional[PythonType]:
+        """Deserialize a flyte literal and return the python object."""
+        if not cls.is_attached():
+            raise Exception("Must attach to a literal map before deserializing")
+        literal = cls.literal_store[identifier]  # type: ignore
+        python_object = deserialize_flyte_literal(literal, expected_type)
+        return python_object
+
+
+def set_validators_on_supported_flyte_types() -> None:
+    """
+    Set pydantic validator for the flyte types supported by this plugin.
+    """
+    for flyte_type in commons.PYDANTIC_SUPPORTED_FLYTE_TYPES:
+        setattr(flyte_type, PYDANTIC_VALIDATOR_METHOD_NAME, add_flyte_validators_for_type(flyte_type))
+
+
+def add_flyte_validators_for_type(
+    flyte_obj_type: Type[type_engine.T],
+) -> Callable[[Any], Iterator[Callable[[Any], type_engine.T]]]:
+    """
+    Add flyte deserialisation validators to a type.
+    """
+
+    previous_validators = getattr(flyte_obj_type, PYDANTIC_VALIDATOR_METHOD_NAME, lambda *_: [])()
+
+    def validator(object_uid_maybe: Union[commons.LiteralObjID, Any]) -> Union[type_engine.T, Any]:
+        """Partial of deserialize_flyte_literal with the object_type fixed"""
+        if not PydanticDeserializationLiteralStore.is_attached():
+            return object_uid_maybe  # this validator should only trigger when we are deserializeing
+        if not isinstance(object_uid_maybe, str):
+            return object_uid_maybe  # object uids are strings and we dont want to trigger on other types
+        if not PydanticDeserializationLiteralStore.contains(object_uid_maybe):
+            return object_uid_maybe  # final safety check to make sure that the object uid is in the literal map
+        return PydanticDeserializationLiteralStore.get_python_object(object_uid_maybe, flyte_obj_type)
+
+    def validator_generator(*args, **kwags) -> Iterator[Callable[[Any], type_engine.T]]:
+        """Generator that returns validators."""
+        yield validator
+        yield from previous_validators
+        yield from ADDITIONAL_FLYTETYPE_VALIDATORS.get(flyte_obj_type, [])
+
+    return validator_generator
+
+
+def validate_flytefile(flytefile: Union[str, file.FlyteFile]) -> file.FlyteFile:
```

Comment:
don't really understand these validators.  is this really required?  is this a pydantic thing?  i feel like we should let the base flytekit type engine handle the transformations